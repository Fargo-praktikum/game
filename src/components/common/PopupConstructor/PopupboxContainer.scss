.popupbox {
    width: 100%;
    height: 100%;
    position: fixed;
    top: 0;
    left: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    opacity: 0;
    pointer-events: none;
    z-index: 1000;
}

.popupbox[hidden] {
    display: none;
}

.popupbox-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1002;
    background: none repeat scroll 0 0 #000;
}

.popupbox-content {
    overflow: hidden;
}

.popupbox-wrapper {
    display: flex;
    flex-direction: column;
    z-index: 1003;
    position: relative;
}

.popupbox-titleBar {
    overflow: hidden;
    display: block;
    position: relative;
}

.popupbox-btn--close {
    z-index: 1004;
}

.popupbox[data-title="bottom"] .popupbox-content {
    order: 1;
}

.popupbox[data-title="bottom"] .popupbox-titleBar {
    order: 2;
}

/* ----- тема по умолчанию ----- */

.popupbox-wrapper {
    border-radius: 5px;
    overflow: hidden;
    /*max-width: 80%;*/
    min-width: 300px;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.9);
    background-color: #fff;
}

.popupbox-content {
    padding: 50px 80px;
}

.popupbox-content_withoutPadding {
    padding: 0;
}

.popupbox-btn--close {
    transition: all 0.2s;
    position: absolute;
    right: 26px;
    top: 26px;
    cursor: pointer;
}

.popupbox-btn--close:hover {
    filter: brightness(70%);
}

.popupbox.is-active {
    opacity: 1;
    pointer-events: auto;
}

.popupbox[data-title="bottom"] .popupbox-content {
    box-shadow: 0 1px 1px rgba(0, 0, 0, 0.3);
}

.popupbox[data-title="bottom"] .popupbox-titleBar {
    box-shadow: none;
    border-top: 1px solid #ccc;
}
