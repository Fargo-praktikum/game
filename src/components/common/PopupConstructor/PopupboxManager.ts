import { EventEmitter } from "events";
import merge from "deepmerge";
import { defaultConfigType, ListenerType } from "./types";

const Constants = {
    OPEN: "open",
    CLOSE: "close",
    CHANGE: "change"
};

type openType = {
    content: JSX.Element,
    config?: Partial<defaultConfigType>,
};


class PopupboxManager extends EventEmitter {
    content: JSX.Element | null = null;
    config: Partial<defaultConfigType> = {};
    show = false;

    private _defaultConfig: defaultConfigType | null = null;

    constructor() {
        super();

        this.open = this.open.bind(this);
        this.update = this.update.bind(this);
        this.close = this.close.bind(this);
    }

    setDefault(defaultConfig: defaultConfigType) {
        this._defaultConfig = defaultConfig;
    }

    open({ content, config = {} }: openType) {
        this.content = content || null;

        this.config = config || this._defaultConfig;

        this.show = true;
        this.emitChange();
    }

    update({ content, config = {} }: openType) {
        this.content = content || this.content;
        this.config = merge(this.config, config);
        this.emitChange();
    }

    close() {
        this.show = false;
        this.emitChange();
    }

    emitChange() {
        this.emit(Constants.CHANGE, {
            children: this.content,
            config: this.config,
            show: this.show
        });
    }

    addChangeListener(callback: ListenerType) {
        this.addListener(Constants.CHANGE, callback);
    }

    removeChangeListener(callback: ListenerType) {
        this.removeListener(Constants.CHANGE, callback);
    }
}

export default new PopupboxManager();
