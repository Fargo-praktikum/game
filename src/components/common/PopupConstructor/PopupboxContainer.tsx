import React, {Component, MouseEventHandler} from 'react'
import Manager from './PopupboxManager'
import merge from 'deepmerge'

import "./PopupboxContainer.scss";

import closePopup from "../../../assets/close.svg";

import {defaultConfigType, getConfigType, handleStoreChangeType} from "./types";


export class PopupboxContainer extends Component<Partial<defaultConfigType>, defaultConfigType> {
    private _defaultState: defaultConfigType;
    closeImagebox: MouseEventHandler<HTMLImageElement>;

    constructor(props: Partial<defaultConfigType>) {
        super(props)

        this._defaultState = this.getConfig({params: props, isInit: true})
        this.state = this._defaultState
        Manager.setDefault(this._defaultState)

        this.handleStoreChange = this.handleStoreChange.bind(this)
        this.closeImagebox = Manager.close.bind(Manager)
    }

    getConfig({params, isInit}: getConfigType) {
        const defaultConfig: defaultConfigType = {
            overlayOpacity: 0.75,
            show: false,
            fadeIn: false,
            fadeInSpeed: 500,
            fadeOut: true,
            fadeOutSpeed: 500,
            overlayClose: true,
            escClose: true,
            closeBtn: true,
            padding: true,
            content: {}
        }

        if (isInit && !params) return defaultConfig

        return merge(isInit ? defaultConfig : this._defaultState, params)
    }

    onKeyDown(e: KeyboardEvent) {
        if (!this.state.escClose) return
        if (this.state.show && (e.code === "Escape")) {
            this.closeImagebox(e as any)
        }
    }

    // componentWillMount() {
    //   Manager.addChangeListener(this.handleStoreChange)
    // }

    componentDidMount() {
        Manager.addChangeListener(this.handleStoreChange);

        document.addEventListener('keydown', this.onKeyDown.bind(this))
    }

    componentWillUnmount() {
        document.removeEventListener('keydown', this.onKeyDown.bind(this))
        Manager.removeChangeListener(this.handleStoreChange)
    }

    handleStoreChange(params: handleStoreChangeType) {
        this.cleanUp()

        const {children, show, config} = params
        const currentConfig = this.getConfig({params: config, isInit: false})
        const {fadeIn, fadeInSpeed, fadeOut, fadeOutSpeed} = currentConfig

        if (show) {
            const {onComplete, onOpen} = currentConfig
            this.setState(merge(currentConfig, {
                children: children,
                show: true,
                transition: (fadeIn) ? `all ${fadeInSpeed / 1000}s ease-in-out` : 'none',
                callback: setTimeout(() => {
                    onComplete && onComplete()
                }, fadeInSpeed + 1)
            }))
            onOpen && onOpen()
        } else {
            const {onCleanUp} = currentConfig
            onCleanUp && onCleanUp()
            this.setState({
                show: false,
                transition: (fadeOut) ? `all ${fadeOutSpeed / 1000}s ease-in-out` : 'none',
                callback: setTimeout(() => {
                    this.onClosed()
                }, fadeOutSpeed + 1)
            })
        }
    }


    onClosed() {
        const {onClosed} = this.state
        onClosed && onClosed()
        this.setState(this._defaultState)
    }

    cleanUp() {
        if (this.state.callback) clearTimeout(this.state.callback)
    }

    render() {
        const {
            overlayOpacity,
            show,
            children,
            style,

            closeBtn,
            padding,
            content
        } = this.state


        return (
            <div
                style={{transition: this.state.transition}}
                className={`popupbox${show ? ' is-active' : ''}`}
            >
                <div
                    className={`popupbox-wrapper`}
                    style={style ? style : undefined}
                >
                    {closeBtn &&
                    <img className={`popupbox-btn--close`}
                         src={closePopup} alt="закрыть" onClick={this.closeImagebox}/>}
                    <div
                        className={`popupbox-content  ${!padding ? 'popupbox-content_withoutPadding' : ''}`}
                        style={content.style ? content.style : undefined}
                    >
                        {children}
                    </div>
                </div>
                <div
                    className="popupbox-overlay"
                    style={{opacity: overlayOpacity}}
                    // style={{ opacity: 0.5 }}
                    onClick={this.state.overlayClose ? this.closeImagebox : undefined}
                />
            </div>
        )
    }
}
