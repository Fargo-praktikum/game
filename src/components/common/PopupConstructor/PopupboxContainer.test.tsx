import React from "react";
import renderer from "react-test-renderer";

import { PopupboxContainer } from "./PopupboxContainer";
import PopupboxManager from "./PopupboxManager";


describe("Проверка PopupboxContainer с помощью snapshot", () => {
    it("корректная отрисовка", () => {
        const tree = renderer.create(
            <PopupboxContainer/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("открытие popup с содержимым", () => {
        const popupContainer = renderer.create(
            <PopupboxContainer/>
        );
        const content = (
            <h1>ПРИВЕТ</h1>
        );
        PopupboxManager.open({
            content,
            config: {
                fadeIn: true,
                fadeInSpeed: 200,
                padding: false
            }
        });
        expect(popupContainer.toJSON()).toMatchSnapshot();
    });

    it("открытие popup с содержимым и обновление", () => {
        const popupContainer = renderer.create(
            <PopupboxContainer/>
        );
        const contentOne = (
            <h1>Первый контент</h1>
        );
        const contentTwo = (
            <h1>Обновленный контент</h1>
        );
        PopupboxManager.open({
            content: contentOne,
            config: {
                fadeIn: true,
                fadeInSpeed: 200,
                padding: false
            }
        });
        PopupboxManager.update({
            content: contentTwo
        });
        expect(popupContainer.toJSON()).toMatchSnapshot();
    });


});



