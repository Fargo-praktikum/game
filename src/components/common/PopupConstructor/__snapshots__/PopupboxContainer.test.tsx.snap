// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Проверка PopupboxContainer с помощью snapshot корректная отрисовка 1`] = `
<div
  className="popupbox"
  style={
    Object {
      "transition": undefined,
    }
  }
>
  <div
    className="popupbox-wrapper"
  >
    <img
      alt="закрыть"
      className="popupbox-btn--close"
      onClick={[Function]}
      src="test-file-stub"
    />
    <div
      className="popupbox-content  "
    />
  </div>
  <div
    className="popupbox-overlay"
    onClick={[Function]}
    style={
      Object {
        "opacity": 0.75,
      }
    }
  />
</div>
`;

exports[`Проверка PopupboxContainer с помощью snapshot открытие popup с содержимым 1`] = `
<div
  className="popupbox is-active"
  style={
    Object {
      "transition": "all 0.2s ease-in-out",
    }
  }
>
  <div
    className="popupbox-wrapper"
  >
    <img
      alt="закрыть"
      className="popupbox-btn--close"
      onClick={[Function]}
      src="test-file-stub"
    />
    <div
      className="popupbox-content  popupbox-content_withoutPadding"
    >
      <h1>
        ПРИВЕТ
      </h1>
    </div>
  </div>
  <div
    className="popupbox-overlay"
    onClick={[Function]}
    style={
      Object {
        "opacity": 0.75,
      }
    }
  />
</div>
`;

exports[`Проверка PopupboxContainer с помощью snapshot открытие popup с содержимым и обновление 1`] = `
<div
  className="popupbox is-active"
  style={
    Object {
      "transition": "all 0.2s ease-in-out",
    }
  }
>
  <div
    className="popupbox-wrapper"
  >
    <img
      alt="закрыть"
      className="popupbox-btn--close"
      onClick={[Function]}
      src="test-file-stub"
    />
    <div
      className="popupbox-content  popupbox-content_withoutPadding"
    >
      <h1>
        Обновленный контент
      </h1>
    </div>
  </div>
  <div
    className="popupbox-overlay"
    onClick={[Function]}
    style={
      Object {
        "opacity": 0.75,
      }
    }
  />
</div>
`;
